clientUtils/SqlFilterSExpression.ts

```diff
@@ -0,0 +1,515 @@
+    [Arity.nary, undefined],
+])
+
+export interface Operation {
+    toSql(): string
+    toSExpr(): string
+    expressionType: ExpressionType
+}
+
+export abstract class NumericOperation implements Operation {
+    abstract toSql(): string
+    abstract toSExpr(): string
+    expressionType: ExpressionType = ExpressionType.numeric
+}
+
+export abstract class BooleanOperation implements Operation {
+    abstract toSql(): string
+    abstract toSExpr(): string
+    expressionType: ExpressionType = ExpressionType.boolean
+}
+
+export abstract class StringOperation implements Operation {
+    abstract toSql(): string
+    abstract toSExpr(): string
+    expressionType: ExpressionType = ExpressionType.string
+}
+
+export enum ExpressionType {
+    numeric = "numeric",
+    boolean = "boolean",
+    string = "string",
+    any = "any", // This is used for values we can't type check e.g. jsonpath queries that might return anything
+}
+
+export class BooleanAtom extends BooleanOperation {
+    constructor(v: boolean) {
+        super()
+        this.value = v
+    }
+    value: boolean
+    toSql(): string {
+        return this.value.toString()
+    }
+
+    toSExpr(): string {
+        return this.value.toString()
+    }
+}
+
+export class NumberAtom extends NumericOperation {
+    constructor(v: number) {
+        super()
+        this.value = v
+    }
+    value: number
+    toSql(): string {
+        return this.value.toString()
+    }
+
+    toSExpr(): string {
+        return this.value.toString()
+    }
+}
+
+export class StringAtom extends StringOperation {
+    constructor(v: string) {
+        super()
+        this.value = v
+    }
+    expressionType = ExpressionType.string
+    value: string
+    escapedValue(): string {
+        return this.value.toString().replace("'", "''")
+    }
+    toSql(): string {
+        return `'${this.escapedValue()}'` // escape single quotes to avoid SQL injection attacks :)
+    }
+
+    toSExpr(): string {
+        return `"${this.value.toString()}"`
+    }
+}
+
+function jsonPointerToJsonPath(jsonpointer: string): string {
+    const pointerFragments = pointer.decode(jsonpointer)
+    let path = "$"
+    for (const fragment of pointerFragments) {
+        const number = Number.parseInt(fragment)
+        if (!Number.isNaN(number)) path += `[${number}]`
+        else path += `.${fragment}`
+    }
+    return path
+}
+
+export class JsonPointerSymbol implements Operation {
+    // Json pointer is actually less restrictive and allows pretty much all unicode
+    // code points but for the foreseeable future we will only have \w and \d, _ and -
+    // characters in actual use. Tilde is from json pointer to escape / inside field names
+    // We also disallow empty jsonPointer for now even though they are legal (if you)
+    // want to enable them then make sure that the parsing decisions outside of here
+    // all correctly allow this)
+    static jsonPointerRegex: RegExp = /^\/[\w\d/-~]+$/
```

Comment:
Be careful with this regex - `/-~` is a range, allowing every ASCII character between `/` and `~`.
Instead, do `\-/~_` (so we also include `_`).

Reply:
Ah that's what you get for not checking again after modifications - I had \d-~ before in which case you don't need to escape the dash. Added your suggestion